datalad/support/annexrepo.py

```diff
@@ -2770,10 +2776,63 @@ def commit(self, msg=None, options=None, _datalad_msg=False,
             #       example.
 
             try:
-                super(AnnexRepo, self).commit(msg, options,
-                                              _datalad_msg=_datalad_msg,
-                                              careless=careless, files=files)
+                alt_index_file = None
+                files_to_commit = None  # we might need to avoid explicit paths
+                if self.is_direct_mode() and files:
+                    # In direct mode, if we commit file(s) they would get
+                    # committed directly into git ignoring possibly being
+                    # staged by annex.  So, if not all files are committed, and
+                    # assuming that what is to be committed is staged (!could be
+                    # wrong assumption), we need to prepare a custom index, and
+                    # commit without specifying any paths
+                    changed_files = {
+                        staged: {
+                            opj(self.path, p)
+                            for p in self.get_changed_files(staged=staged)
+                        } for staged in (True,) #  False}
+                    }
+                    files_set = set(files)
+                    files_notstaged = files_set.difference(changed_files[True])
+                    # Lazy .add('.') results in ALL (performance?) files listed
+                    # here even if they are not changed at all and could be ignored
+                    #files_changed_notstaged = files_notstaged.difference(changed_files[False])
+                    # if files_changed_notstaged:
+                    #     # To be safe(r), let's just blow for now!
+                    #     # We could be smart(er) and do all the git or annex add
+                    #     # but let's see if we run into those cases first
+                    #     import pdb; pdb.set_trace()
+                    #     raise RuntimeError(
+                    #         "To commit files in direct mode, please first git "
+                    #         "or git-annex add them first!  Files: %s"
+                    #         % ', '.join(files_changed_notstaged)
+                    #     )
+                    # Files which were staged but not among files
+                    staged_not_to_commit = changed_files[True].difference(files_set)
+                    if staged_not_to_commit:
+                        # Need an alternative index_file
+                        with make_tempfile() as index_file:
+                            alt_index_file = index_file
+                            # Reset the files we are not to be committed
+                            self._git_custom_command(
+                                list(staged_not_to_commit),
+                                ['git', 'reset'],
+                                index_file=alt_index_file)
```

Comment:
Huh I expected a write-tree/read-tree call somewhere, or a copy on the current index.  AFAICS what you end up with is an index that only has the files in `staged_not_to_commit` (at HEAD's version).  IOW, you have all files in the working tree except for those in `staged_not_to_commit` staged for deletion, and modifications in `staged_not_to_commit` as unstaged changes.  Once you go back to .git/index, you'd have the modifications from `staged_not_to_commit` staged, but you'd also have "A"s for all the other files in the working tree.

Reply:
As of cc0bcd663e73, this copies the current index.